File: util/src/main/java/etch/util/AbstractStartable.java
/* $Id$
 *
 * Copyright 2007-2008 Cisco Systems Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy
 * of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

package etch.util;

/**
 * Generic implementation of Startable.
 */
abstract public class AbstractStartable implements Startable
{
	final public synchronized void start() throws Exception
	{
		if (isStarted())
			throw new IllegalStateException( "is already started" );
		
		try
		{
			started = true;
			start0();
		}
		catch ( Exception e )
		{
			setStopped();
			stop0();
			throw e;
		}
	}

	final public synchronized void stop() throws Exception
	{
		checkIsStarted();
		setStopped();
		stop0();
	}
	
	/**
	 * Signals (from a subclasser) that the startable has stopped.
	 */
	final protected void setStopped()
	{
		started = false;
	}
	
	final public boolean isStarted()
	{
		return started;
	}
	
	final public void checkIsStarted()
	{
		if (!isStarted())
			throw new IllegalStateException( "is not started" );
	}

	private boolean started;
	
	/**
	 * @throws Exception
	 */
	abstract protected void start0() throws Exception;
	
	/**
	 * @throws Exception
	 */
	abstract protected void stop0() throws Exception;
}

File: binding-csharp/runtime/src/main/csharp/Etch/Util/AbstractStartable.cs
// $Id$
// 
// Copyright 2007-2008 Cisco Systems Inc.
// 
// Licensed under the Apache License, Version 2.0 (the "License"); you may not
// use this file except in compliance with the License. You may obtain a copy
// of the License at
//  
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations
// under the License.

using System;
using System.Runtime.CompilerServices;

namespace Etch.Util
{
    /// <summary>
    /// Generic implementation of Startable
    /// </summary>
    abstract public class AbstractStartable : Startable
    {
        public void Start()
        {
            lock (this)
            {
                if (IsStarted())
                    throw new Exception("is already started");
                SetStarted();
            }

            try
            {
                Start0();
            }
            catch (Exception e)
            {
                Stop();
                throw;
            }
        }

        public void Stop()
        {
            lock (this)
            {
                CheckIsStarted();
                SetStopped();
            }

            Stop0();
        }

        private bool started;

        private void SetStarted()
        {
            started = true;
        }

        protected void SetStopped()
        {
            started = false;
        }

        public bool IsStarted()
        {
            return started;
        }

        /// <summary>
        /// Check that the startable is started
        /// Exception:
        ///     ThreadInterruptedException if not started
        /// </summary>
        /// 
        public void CheckIsStarted()
        {
            if (!IsStarted())
                throw new Exception("is not started");
        }

        /// <summary>
        /// 
        /// </summary>
        /// Exception:
        ///     throws Exception
        abstract protected void Start0();

        /// <summary>
        /// 
        /// </summary>
        /// Exception:
        ///     throws Exception
        abstract protected void Stop0();

    }
}
